utils/filesystem.h

```diff
@@ -0,0 +1,125 @@
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#ifndef UTILS_FILESYSTEM_H
+#define UTILS_FILESYSTEM_H
+
+#include <stdbool.h>
+#include <sys/types.h>
+
+/*
+ * Check if the given path exist
+ * */
+bool pv_filesystem_path_exist(const char *path);
+/*
+ * check if the given path is a directory
+ * */
+bool pv_filesystem_path_is_directory(const char *path);
+/*
+ * Sync the given path. If path points to a file,
+ * then the parent folder is synced
+ * */
+void pv_filesystem_path_sync(const char *path);
+/*
+ * Path concatenation, the value is returned on the given buffer
+ * */
+void pv_filesystem_path_join(char *buf, int size, ...);
+/*
+ * Creates a directory with the given path and mode
+ * All subdirectories will be created.
+ * */
+int pv_filesystem_mkdir_p(const char *path, mode_t mode);
+/*
+ * Remove a path (file or directory). If recursive is true then all
+ * subdirectories will be deleted
+ * */
+int pv_filesystem_path_remove(const char *path, bool recursive);
+/*
+ * Rename a path(file or directory)
+ * */
+int pv_filesystem_path_rename(const char *src_path, const char *dst_path);
+/*
+ * Returns the file size for the given path
+ * */
+size_t pv_filesystem_path_get_size(const char *path);
+/*
+ * Creates a temporary path based on the given fname
+ * */
+int pv_filesystem_file_tmp(char *tmp, const char *fname);
+/* Save the given data on fname using a specific mode.
+ * If the file doesn't exist, will be created.
+ * If file exists, it will be truncated.
+ * */
+int pv_filesystem_file_save(const char *fname, const char *data, mode_t mode);
+/*
+ * Copy a file from src to dst.
+ * */
+int pv_filesystem_file_copy_from_path(const char *src, const char *dst, mode_t mode);
+/*
+ * Copy the contents of the src file descriptor to dst. If close_src is true
+ * the file pointed by src will be closed.
+ * */
+ssize_t pv_filesystem_file_copy_from_fd(int src, int dst, bool close_src);
+/*
+ * Get an extended attribute from the given file.
+ * NOTE: This function allocate a new buffer and it can be freed with free()
+ * */
+char *pv_filesystem_file_get_xattr_dup(const char *fname, const char *attr);
```

Comment:
why pv_filesystem_file_get_xattr_dup and pv_filesystem_file_get_xattr? if we have to decide between the two, I'd prefer the second maybe?

Reply:
Yeah, the difference is one receives a buffer and the other allocates a buffer for you.
In the old library (file.h), this function receives a pointer to function to allocate memory on the value field but is it always called with NULL and a previously allocated buffer. The problem with that is `getxattr` needs the size of the attribute value. The documentation says that you can call `getxattr` with `size = 0` to retrieve the size but if the buffer was allocated before the size can be smaller than it's needed.
This is the code in `file.h`

```c
int pv_file_get_file_xattr(const char *filename, char *attr, char **dst, int(*alloc)(char**, int))
{
  int val_len = -1;

  val_len = getxattr(filename, attr, NULL, 0);
  if (val_len > 0) {
    int ret = 0;

    // if you use a NULL pointer this never happen
    if (alloc)
      ret = alloc(dst, val_len + 1);
    // even with a NULL pointer ret = 0 
    if (!ret)
      val_len = getxattr(filename, attr, *dst, val_len);  // var_len was come from the previous getxattr call
  }                                                                              // so that value is not related with dst.  
  return val_len < 0 ? -errno : val_len;
}

```
Maybe, as you said, the second option is the best. If we are OK, I can remove the first one

Reply:
yeah, let's do that. Then we can see how to fit those in the current code in the next PR

Reply:
Removed in 1616ea96eca28383bb3904e4f009bdba3ad23d75